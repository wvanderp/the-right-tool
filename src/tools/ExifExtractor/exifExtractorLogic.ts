import { parseMetadata } from "@uswriting/exiftool";

// Define a type for the expected EXIF data structure
export type ExifData = Record<string, unknown>[];

/**
 * Extracts EXIF metadata from an image file.
 * @param file The image file to process.
 * @returns A promise that resolves with the extracted EXIF data or an error message.
 */
export async function extractExifDataFromFile(
    file: File
): Promise<{ data: ExifData | null; error: string | null }> {
    try {
        // Specify ExifData as the expected type T for the 'data' field in the result
        const exifResult = await parseMetadata<ExifData>(file, {
            args: ["-json"], // Ask exiftool to output JSON
            // Transform the raw string output (which should be JSON) into the ExifData type
            transform: (data: string): ExifData => JSON.parse(data),
        });

        // Check the success flag from the exifResult object
        if (exifResult.success && exifResult.data) {
            // Return the actual data payload
            return { data: exifResult.data, error: null };
        } else {
            // Use the error message from the exifResult object
            return {
                data: null,
                error: exifResult.error || "Failed to extract EXIF data.",
            };
        }
    } catch (err) {
        if (err instanceof Error) {
            return {
                data: null,
                error: err.message || "Failed to extract EXIF data.",
            };
        } else {
            return {
                data: null,
                error: "An unknown error occurred during EXIF extraction.",
            };
        }
    }
}
