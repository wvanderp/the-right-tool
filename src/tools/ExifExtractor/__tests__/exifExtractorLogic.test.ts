import { describe, it, expect } from "vitest";
import { extractExifDataFromFile } from "../exifExtractorLogic";
import fs from "fs";
import path from "path";

describe("ExifExtractor Logic", () => {
    it("should correctly extract EXIF data from a sample image", async () => {
        // Create a File object from the test image
        const imagePath = path.join(
            __dirname,
            "2023-08-04_Saemangeum_temp_phone_mast_with_climber_-_WSJ_2023.jpg"
        );
        const imageBuffer = fs.readFileSync(imagePath);
        const imageFile = new File(
            [imageBuffer],
            "2023-08-04_Saemangeum_temp_phone_mast_with_climber_-_WSJ_2023.jpg",
            { type: "image/jpeg" }
        );

        // Extract EXIF data from the image
        const { data, error } = await extractExifDataFromFile(imageFile);

        // Check that there is no error
        expect(error).toBeNull();

        // Check that data is not null
        expect(data).not.toBeNull();

        // Make sure we have EXIF data in the expected format
        expect(Array.isArray(data)).toBeTruthy();
        expect(data!.length).toBeGreaterThan(0);

        // Check for some common EXIF properties that should exist
        const exifData = data![0];
        expect(exifData).toHaveProperty("FileName");
        expect(exifData.FileName).toBe(
            "2023-08-04_Saemangeum_temp_phone_mast_with_climber_-_WSJ_2023.jpg"
        );

        // Take a snapshot of the entire EXIF data structure
        // This will create a snapshot file to compare against in future test runs
        expect(data).toMatchSnapshot();
    });

    it("should handle errors when file is not an image", async () => {
        // Create a non-image file
        const textFile = new File(["This is not an image"], "test.txt", {
            type: "text/plain",
        });

        // Try to extract EXIF data
        const { data } = await extractExifDataFromFile(textFile);

        // Should return file-level metadata for a text file
        expect(Array.isArray(data)).toBeTruthy();
        expect(data).toHaveLength(1);
        expect(data?.[0].FileType).toBe("TXT");
        expect(data?.[0].FileName).toBe("test.txt");
        // Error may be null or a warning, depending on exiftool's output
    });
});
