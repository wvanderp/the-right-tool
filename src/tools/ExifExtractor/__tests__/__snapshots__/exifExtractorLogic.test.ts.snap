// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExifExtractor Logic > should correctly extract EXIF data from a sample image 1`] = `
[
  {
    "Aperture": 3.5,
    "ApertureValue": 3.5,
    "BitsPerSample": 8,
    "BlueMatrixColumn": "0.14307 0.06061 0.7141",
    "BlueTRC": "(Binary data 2060 bytes, use -b option to extract)",
    "BrightnessValue": 8.59,
    "CMMFlags": "Not Embedded, Independent",
    "ChromaticAdaptation": "1.04791 0.02293 -0.0502 0.0296 0.99046 -0.01707 -0.00925 0.01506 0.75179",
    "CircleOfConfusion": "0.005 mm",
    "ColorComponents": 3,
    "ColorSpace": "sRGB",
    "ColorSpaceData": "RGB ",
    "ComponentsConfiguration": "Y, Cb, Cr, -",
    "CompositeImage": "Composite Image Captured While Shooting",
    "Compression": "JPEG (old-style)",
    "ConnectionSpaceIlluminant": "0.9642 1 0.82491",
    "Contrast": "Normal",
    "CreateDate": "2023:08:04 18:37:17",
    "CustomRendered": "Custom",
    "DOF": "inf (21.14 m - inf)",
    "DateTimeOriginal": "2023:08:04 18:37:17",
    "DeviceAttributes": "Reflective, Glossy, Positive, Color",
    "DeviceManufacturer": "",
    "DeviceModel": "",
    "DeviceModelDesc": "IEC 61966-2-1 Default RGB Colour Space - sRGB",
    "DigitalZoomRatio": 1.03,
    "Directory": "/",
    "DirectoryItemLength": 0,
    "DirectoryItemMime": "image/jpeg",
    "DirectoryItemPadding": 0,
    "DirectoryItemSemantic": "Primary",
    "EncodingProcess": "Baseline DCT, Huffman coding",
    "ExifByteOrder": "Little-endian (Intel, II)",
    "ExifImageHeight": 3853,
    "ExifImageWidth": 2167,
    "ExifToolVersion": 13.3,
    "ExifVersion": "0232",
    "ExposureCompensation": 0,
    "ExposureMode": "Auto",
    "ExposureProgram": "Program AE",
    "ExposureTime": "1/231",
    "FNumber": 3.5,
    "FOV": "17.5 deg",
    "FileAccessDate": "0000:00:00 00:00:00",
    "FileInodeChangeDate": "1970:01:01 00:00:42+00:00",
    "FileModifyDate": "0000:00:00 00:00:00",
    "FileName": "2023-08-04_Saemangeum_temp_phone_mast_with_climber_-_WSJ_2023.jpg",
    "FilePermissions": "----------",
    "FileSize": "541 kB",
    "FileType": "JPEG",
    "FileTypeExtension": "jpg",
    "Flash": "Off, Did not fire",
    "FlashpixVersion": "0100",
    "FocalLength": "19.0 mm",
    "FocalLength35efl": "19.0 mm (35 mm equivalent: 117.0 mm)",
    "FocalLengthIn35mmFormat": "117 mm",
    "GPSAltitude": "36.8 m Above Sea Level",
    "GPSAltitudeRef": "Above Sea Level",
    "GPSDateStamp": "2023:08:04",
    "GPSDateTime": "2023:08:04 09:36:43Z",
    "GPSImgDirection": 227,
    "GPSImgDirectionRef": "Magnetic North",
    "GPSLatitude": "35 deg 42' 40.94" N",
    "GPSLatitudeRef": "North",
    "GPSLongitude": "126 deg 35' 46.62" E",
    "GPSLongitudeRef": "East",
    "GPSPosition": "35 deg 42' 40.94" N, 126 deg 35' 46.62" E",
    "GPSTimeStamp": "09:36:43",
    "GPSVersionID": "2.2.0.0",
    "GreenMatrixColumn": "0.38515 0.71687 0.09708",
    "GreenTRC": "(Binary data 2060 bytes, use -b option to extract)",
    "HyperfocalDistance": "21.14 m",
    "ISO": 23,
    "ImageHeight": 3853,
    "ImageSize": "2167x3853",
    "ImageUniqueID": "5b290bcfda2a3cf80000000000000000",
    "ImageWidth": 2167,
    "InteropIndex": "R98 - DCF basic file (sRGB)",
    "InteropVersion": "0100",
    "JFIFVersion": 1.01,
    "LensID": "Pixel 7 Pro back camera 19.0mm f/3.5",
    "LensMake": "Google",
    "LensModel": "Pixel 7 Pro back camera 19.0mm f/3.5",
    "LightValue": 13.6,
    "Luminance": "0 80 0",
    "MIMEType": "image/jpeg",
    "Make": "Google",
    "MaxApertureValue": 3.5,
    "MeasurementBacking": "0 0 0",
    "MeasurementFlare": "0%",
    "MeasurementGeometry": "Unknown",
    "MeasurementIlluminant": "D65",
    "MeasurementObserver": "CIE 1931",
    "MediaBlackPoint": "0.01205 0.0125 0.01031",
    "MediaWhitePoint": "0.9642 1 0.82491",
    "Megapixels": 8.3,
    "MeteringMode": "Center-weighted average",
    "Model": "Pixel 7 Pro",
    "ModifyDate": "2023:08:04 18:37:17",
    "OffsetTime": "+09:00",
    "OffsetTimeDigitized": "+09:00",
    "OffsetTimeOriginal": "+09:00",
    "Orientation": "Horizontal (normal)",
    "PrimaryPlatform": "Unknown ()",
    "ProfileCMMType": "",
    "ProfileClass": "Display Device Profile",
    "ProfileConnectionSpace": "XYZ ",
    "ProfileCopyright": "Copyright International Color Consortium, 2009",
    "ProfileCreator": "",
    "ProfileDateTime": "2009:03:27 21:36:31",
    "ProfileDescription": "sRGB IEC61966-2-1 black scaled",
    "ProfileFileSignature": "acsp",
    "ProfileID": "29f83ddeaff255ae7842fae4ca83390d",
    "ProfileVersion": "2.0.0",
    "RedMatrixColumn": "0.43607 0.22249 0.01392",
    "RedTRC": "(Binary data 2060 bytes, use -b option to extract)",
    "RenderingIntent": "Perceptual",
    "ResolutionUnit": "inches",
    "Saturation": "Normal",
    "ScaleFactor35efl": 6.2,
    "SceneCaptureType": "Standard",
    "SceneType": "Directly photographed",
    "SensingMethod": "One-chip color area",
    "Sharpness": "Normal",
    "ShutterSpeed": "1/231",
    "ShutterSpeedValue": "1/231",
    "Software": "HDR+ 1.0.540104767zd",
    "SourceFile": "/2023-08-04_Saemangeum_temp_phone_mast_with_climber_-_WSJ_2023.jpg",
    "SubSecCreateDate": "2023:08:04 18:37:17.854+09:00",
    "SubSecDateTimeOriginal": "2023:08:04 18:37:17.854+09:00",
    "SubSecModifyDate": "2023:08:04 18:37:17.854+09:00",
    "SubSecTime": 854,
    "SubSecTimeDigitized": 854,
    "SubSecTimeOriginal": 854,
    "SubjectDistance": "4294967295 m",
    "SubjectDistanceRange": "Distant",
    "Technology": "Cathode Ray Tube Display",
    "ThumbnailImage": "(Binary data 2831 bytes, use -b option to extract)",
    "ThumbnailLength": 2831,
    "ThumbnailOffset": 1402,
    "ViewingCondDesc": "Reference Viewing Condition in IEC 61966-2-1",
    "Warning": "Ignored non-standard extended XMP (GUID DE0F81822D030BDD73E5747DAD7A1C6A)",
    "WhiteBalance": "Auto",
    "XResolution": 72,
    "YCbCrPositioning": "Centered",
    "YCbCrSubSampling": "YCbCr4:2:0 (2 2)",
    "YResolution": 72,
  },
]
`;
